### DISCUSSION TITLE
VRPTW w/ pickups and deliveries problem: idle before next pickup

### DISCUSSION THREAD

**Discussion by odashevsky** [google/or-tools#2823] [Routing (and legacy CP) questions]
Hello,

I would appreciate some guidance for the following problem. I am solving a VRPTW where a number of vehicles goes around town picking up and dropping people. Pickup times (and time to drop-off) is known and some slack when the pickup and drop-off can happen (+/- 15 minutes for the example sake). There's also a capacity constraint here (a vehicle can have 2 riders at a time) but I do not believe its relevant here. Here's the issue. If I have, say, 67 trips and 18 vehicles the solver is failing to find a solution. If I allow to drop trips, it does find a solution, but 30% of trips are dropped. I've noticed the following scenario in one of such solutions: 
1. there's a trip with pickup at location A at 4 am. Drop off is at location B with travel time of 30 minutes (i.e. estimated drop-off 4.30am)
2. there's a trip with pickup at location C at 9 am. Drop off is at location D with travel time of 15 minutes (i.e. estimated drop-off 9.15am)

The solver would give a vehicle a SINGLE trip C -> D and no other trips for the entire day (the day is 24 hours) and would drop the A -> B trip. (Other vehicles are busier but still many trips are dropped though I can see not all vehicles are "busy" or at full capacity throughout the day).

I suspect that the solver does not have a concept of "Idle". Is it possible that the solver thinks that if vehicle V gets the A->B trip it will have nowhere to go after that (C->D opens only ~4 hours later) and thus drops the A->B trip and selects the C->D trip (because its "cheaper")? If so, how do I tell the solver to allow to "idle" (do nothing, sit at B until going to C to be there at 9, or sit at C until C opens at 9)? Obviously, we want vehicles to idle as rarely as possible, but if that's the only way not to drop actual trips then  idling is allowed.

Thank you,
Oleg D.

**Comment by Mizux** [comment]
1. Idle is usually modelised by the slack var so to limit the maximum slack aka maximum waiting time simply limit the max value of the slack var.

2. if each time a node is dropped you have a high penalty then solver should avoid to drop too many node.

3. Supposing solver make the wrong assumption for the initial solution, did you enable the local search so solver can search for a better one ?

**Reply by odashevsky** [reply]
Hi, thanks for answering so quickly.
1. Please forgive my ignorance, I am a noob with or-tools. When you refer to "limit max value of slack var", do you refer to the parameter "slack_max" of the function AddDimension (which I use to define the travel times between locations)? If yes, its currently set to 24 hours (1440 minutes to be exact)  (which I think means "you can wait as long as you can at each location"). Am I reading it wrong?
2. I set penalty to long.MaxValue - 1! Still drops them.
3.  I use local search = GuidedLocalSearch, FirstSolutionStrategy = PathCheapestArc (or PathMostConstrainedArc which produces the same results). If I disallow drops, the solver runs indefinitely. Last time I ran it I killed it after an hour (!). Running that long would not be acceptable anyway even if it did find a solution. Solution does exist - schedule was done manually, I am automating manual labor.

**Reply by Mizux** [reply]
1. YES
2. usually first heuristic are greedy so it can be expected. ok to make it the dominant factor but `1_000_000` should be OK, if using  `long.MaxValue - 1` and two nodes are dropped I fear of overflow...
3. Did you use a time limit ?
https://github.com/google/or-tools/blob/86d4c543f717fa8716a9d66c25143660668bf825/ortools/constraint_solver/samples/vrp_with_time_limit.py#L99-L102
note: You can enable trace (i.e. by using logSearch) to see if the objective drop (aka solver manage to integrate more node)

**Reply by odashevsky** [reply]
Yes, I've tried all of that. When trips drop is not allowed, the solver never finishes or times out (I had it run for hours) and the log search shows no found solutions. When I allow to drop visits with cost 1000000, it finds 1000 solutions (I set a max to 1000 solutions) fairly quickly (seconds) but all solutions drop trips and I can see that some dropped visits (4am ones) could have been given to a vehicle that has a first trip at 9 am.

Please let me make sure that I set my time constraints right. My "day" is 24 hours long (60 * 24 = 1440 minutes). "Depot's" time window is 24 hours: 
timeDimension.CumulVar(manager.NodeToIndex(0)).SetRange(0, 60 * 24); // depot is node 0

The travel Time Matrix between nodes is set in terms of minutes, e.g.:
0   0   0   0   0   0   0   0
0   0   10 20 30 20 10 5
0  10   0 
0  20      
0  30         
0  20              
0  10                 
0  5                     
(above matrix is missing numbers, just don't want to type them, but hopefully you get the idea. Its the same as in all examples and samples). First row and column are all 0s because there's no depot. Vehicle can start anywhere. The function that returns trip cost return the number from the matrix + 10 minutes (10 minutes it takes to load a person, could be a wheelchair).

The call to add time matrix looks like this:
routingModel.AddDimension(transitCallbackIndex, 24 * 60, 24 * 60 * NumberOfVehicles, false, "Time");
As you can see I am giving vehicles all the slack in the world. Or at least I think I am.


Now, the pickup and drop off windows are set in terms of minutes in the day:
Pickup location A: scheduled pickup is 4 am +/- 15 minutes. 
Drop-off location B: travel time: 30 minutes, so scheduled drop off is between 3.45 am + 30 minutes to 4.15 + 30 minutes (I also add more slack there of another 15 minutes for delays, but that's irrelevant for my question).
So, for these two nodes I call:
Location A: timeDimension.CumulVar(indexOfLocationA).SetRange(225, 255); // <--- 3.45 am - 4.15 am in terms of minutes
Location B: timeDimension.CumulVar(indexOfLocationB).SetRange(225, 300); // <--- 3.45 am - 5 am in terms of minutes
**Is that right?**

Then I call the functions necessary to setup pickup and delivery constraints using approaches from examples and samples. I.e. to make sure pickup happens at A and drop-off at B in the example above.
(I do NOT call RoutingModel.AddPickupAndDeliverySets() for pickup and drop-off indices returned by AddDisjunction for pickup and drop-off nodes. **Or do I need to?**)

Then I call the functions necessary to setup capacity constraints (2 riders max per vehicle).

Is my approach sound? 
Thanks.
(BTW I use C# if that makes a difference)
